diff --git a/src/python.c b/src/python.c
index fd8604de..2f9fd803 100644
--- a/src/python.c
+++ b/src/python.c
@@ -931,7 +931,12 @@ execute(ForeignScanState *node, ExplainState *es)
 				newqual->base.opname = qual->opname;
 				newqual->base.isArray = qual->isArray;
 				newqual->base.useOr = qual->useOr;
+
+				#if PG_VERSION_NUM >= 100000
+				newqual->value = ExecEvalExpr(expr_state, econtext, &isNull);
+				#else
 				newqual->value = ExecEvalExpr(expr_state, econtext, &isNull, NULL);
+				#endif
 				newqual->base.typeoid = ((Param*) ((MulticornParamQual *) qual)->expr)->paramtype;
 				newqual->isnull = isNull;
 				break;
