diff -r 984fad7bef53 tox/_venv.py
--- a/tox/_venv.py	Sun Jun 23 09:27:06 2013 +0200
+++ b/tox/_venv.py	Mon Jun 24 17:18:36 2013 +0200
@@ -179,10 +179,6 @@
         venvscript = path.rstrip("co")
         #venvscript = py.path.local(tox.__file__).dirpath("virtualenv.py")
         args = [config_interpreter, venvscript]
-        if self.envconfig.distribute:
-            args.append("--distribute")
-        else:
-            args.append("--setuptools")
         if self.envconfig.sitepackages:
             args.append('--system-site-packages')
         #if sys.platform == "win32":
