--- a/tox/_venv.py	Thu Feb 02 20:10:54 2012 -0500
+++ b/tox/_venv.py	Mon Feb 20 14:31:34 2012 +0100
@@ -163,7 +163,9 @@
         args = [config_interpreter, venvscript]
         if not self._ispython3() and self.envconfig.distribute:
             args.append('--distribute')
-        if not self.envconfig.sitepackages:
+        if self.envconfig.sitepackages:
+            args.append('--system-site-packages')
+        else:
             args.append('--no-site-packages')
         #if sys.platform == "win32":
         #    f, path, _ = py.std.imp.find_module("virtualenv")
